Upgrade EUI to v94.1.0 (major `EuiTable` refactors)

cee-chen opened the pull request:
`v93.6.0` ⏩ `v94.1.0`

> [!important]
> 👋 Hello everyone - this is a special release containing `EuiTable`'s conversion to Emotion, several long-overdue cleanups and breaking changes, and one or two fun new features. First, let's address the big questions:

### Q: I'm listed as a codeowner, how much should I manually QA/review?

Answer: It depends on what exactly in your code changed, but _in general_ I would strongly suggest at least pulling this branch down and doing a quick visual smoke test of all tables (_note: **not** datagrids_) in your apps/plugins. You should not expect to see any major visual regressions.

If your table contained any kind of custom styling or behavior (e.g. custom CSS, etc.) I **strongly** recommend taking more time to QA thoroughly to ensure your table still looks and behaves as expected. Teams with very manual or specific updates will be flagged below in comment threads.

### Q: When do I need to review by?

This PR will be merged **after** 8.14FF. Because this upgrade touches so many files and teams, we're aiming for asking for an admin merge by EOD 4/18 regardless of full approval status.

As always, you're welcome to ping us after merge if you find any issues later ([see our FAQ](https://github.com/elastic/eui/blob/main/wiki/eui-team-processes/upgrading-kibana.md#faq-for-kibana-teams)), as you will have until 8.15FF to catch any bugs.

### Q: What breaking changes were made, and why?

Here's a quick shortlist of all the changes made that affected the majority of the commits in this PR:

#### <u>Removed several top-level table props</u>
- The `responsive` prop has been removed in favor of the new `responsiveBreakpoint` prop (same `false` behavior as before)
- The following props were removed from basic and in-memory tables:
  - `hasActions`, `isSelectable`, and `isExpandable`
    - These props were not used for functionality and were only used for styling tables in mobile/responsive views, which is not a best practice pattern we wanted for our APIs. Mobile tables are now styled correctly without needing consumers to pass these extra props.
  - `textOnly`
    - This prop was unused and had no meaningful impact on tables or table content.

#### Removed hidden mobile vs. desktop DOM

Previously, EUI would set classes that applied `display: none` CSS for content that was hidden for mobile vs. desktop. This is no longer the case, and content that only displays for mobile or only displays for desktop will no longer render to the DOM at all if the table is not in that responsive state.

This is both more performant when rendering large quantities of cells/content, and simpler to write test assertions for when comparing what the user actually sees vs. what the DOM ‘sees’. (c3eeb08441e4b6efe6505e7cddaa87b540ddb259, 78cefcd954a7b46eaccd05e431b5e24dc86071a3)

#### Removed direct usages of table `className`s

EuiTable `classNames` no longer have any styles attached to them, so some instances where Kibana usages were applying the `className` for styles have been replaced with direct component usage or removed entirely (86ce80b61f92137df761d06a1d5f492d9f738dd7).

#### Custom table cell styles

Any custom CSS for table cells was previously being applied to the inner `div.euiTableCellContent` wrapper. As of this latest release, the `className` and `css` props will now be applied directly to the outer `td.euiTableRowCell` element. If you were targeting custom styles table cells, please re-QA your styles to ensure everything still looks as expected.

---

<details open><summary>Full changelog (click to collapse)</summary>

## [`v94.1.0-backport.0`](https://github.com/elastic/eui/releases/v94.1.0-backport.0)

**This is a backport release only intended for use by Kibana.**

**Bug fixes**

- Fixed a visual text alignment regression in `EuiTableRowCell`s with the `row` header scope ([#7681](https://github.com/elastic/eui/pull/7681))

**Accessibility**

- Improved `EuiBasicTable` and `EuiInMemoryTable`'s selection checkboxes to have unique aria-labels per row ([#7672](https://github.com/elastic/eui/pull/7672))

## [`v94.1.0`](https://github.com/elastic/eui/releases/v94.1.0)

- Updated `EuiTableHeaderCell` to show a subdued `sortable` icon for columns that are not currently sorted but can be ([#7656](https://github.com/elastic/eui/pull/7656))
- Updated `EuiBasicTable` and `EuiInMemoryTable`'s `columns[].actions[]`'s to pass back click events to `onClick` callbacks as the second callback ([#7667](https://github.com/elastic/eui/pull/7667))

## [`v94.0.0`](https://github.com/elastic/eui/releases/v94.0.0)

- Updated `EuiTable`, `EuiBasicTable`, and `EuiInMemoryTable` with a new `responsiveBreakpoint` prop, which allows customizing the point at which the table collapses into a mobile-friendly view with cards ([#7625](https://github.com/elastic/eui/pull/7625))
- Updated `EuiProvider`'s `componentDefaults` prop to allow configuring `EuiTable.responsiveBreakpoint` ([#7625](https://github.com/elastic/eui/pull/7625))

**Bug fixes**

- `EuiBasicTable` & `EuiInMemoryTable` `isPrimary` actions are now correctly shown on mobile views ([#7640](https://github.com/elastic/eui/pull/7640))
- Table `mobileOptions`: ([#7642](https://github.com/elastic/eui/pull/7642))
  - `mobileOptions.align` is now respected instead of all cells being forced to left alignment
  - `textTruncate` and `textOnly` are now respected even if a `render` function is not passed

**Breaking changes**

- Removed unused `EuiTableHeaderButton` component ([#7621](https://github.com/elastic/eui/pull/7621))
- Removed the `responsive` prop from `EuiTable`, `EuiBasicTable`, and `EuiInMemoryTable`. Use the new `responsiveBreakpoint` prop instead ([#7625](https://github.com/elastic/eui/pull/7625))
- The following props are no longer needed by `EuiBasicTable` or `EuiInMemoryTable` for responsive table behavior to work correctly, and can be removed: ([#7632](https://github.com/elastic/eui/pull/7632))
  - `isSelectable`
  - `isExpandable`
  - `hasActions`
- Removed the `showOnHover` prop from `EuiTableRowCell` / `EuiBasicTable`/`EuiInMemoryTable`'s `columns` API. Use the new actions `columns[].actions[].showOnHover` API instead. ([#7640](https://github.com/elastic/eui/pull/7640))
- Removed top-level `textOnly` prop from `EuiBasicTable` and `EuiInMemoryTable`. Use `columns[].textOnly` instead. ([#7642](https://github.com/elastic/eui/pull/7642))

**DOM changes**

- `EuiTable` mobile headers no longer render in the DOM when not visible (previously rendered with `display: none`). This may affect DOM testing assertions. ([#7625](https://github.com/elastic/eui/pull/7625))
- `EuiTableRowCell` now applies passed `className`s to the parent `<td>` element, instead of to the inner cell content `<div>`. ([#7631](https://github.com/elastic/eui/pull/7631))
- `EuiTableRow`s rendered by basic and memory tables now only render a `.euiTableRow-isSelectable` className if the selection checkbox is not disabled ([#7632](https://github.com/elastic/eui/pull/7632))
- `EuiTableRowCell`s with `textOnly` set to `false` will no longer attempt to apply the `.euiTableCellContent__text` className to child elements. ([#7641](https://github.com/elastic/eui/pull/7641))
- `EuiTableRowCell` no longer renders mobile headers to the DOM unless the current table is displaying its responsive view. ([#7642](https://github.com/elastic/eui/pull/7642))
- `EuiTableHeaderCell` and `EuiTableRowCell` will no longer render in the DOM at all on mobile if their columns' `mobileOptions.show` is set to `false`. ([#7642](https://github.com/elastic/eui/pull/7642))
- `EuiTableHeaderCell` and `EuiTableRowCell` will no longer render in the DOM at all on desktop if their columns' `mobileOptions.only` is set to `true`. ([#7642](https://github.com/elastic/eui/pull/7642))

**CSS-in-JS conversions**

- Converted `EuiTable`, `EuiTableRow`, `EuiTableRowCell`, and all other table subcomponents to Emotion ([#7654](https://github.com/elastic/eui/pull/7654))
- Removed the following `EuiTable` Sass variables: ([#7654](https://github.com/elastic/eui/pull/7654))
  - `$euiTableCellContentPadding`
  - `$euiTableCellContentPaddingCompressed`
  - `$euiTableCellCheckboxWidth`
  - `$euiTableHoverColor`
  - `$euiTableSelectedColor`
  - `$euiTableHoverSelectedColor`
  - `$euiTableActionsBorderColor`
  - `$euiTableHoverClickableColor`
  - `$euiTableFocusClickableColor`
- Removed the following `EuiTable` Sass mixins: ([#7654](https://github.com/elastic/eui/pull/7654))
  - `euiTableActionsBackgroundMobile`
  - `euiTableCellCheckbox`
  - `euiTableCell`
</details>

cee-chen committed:
Upgrade EUI to v94.1.0

cee-chen committed:
Convert `responsive` prop to new `responsiveBreakpoint` API

- same `false` behavior, but more extensible if needed

cee-chen committed:
Remove unused `textOnly` table prop

cee-chen committed:
Remove `hasActions` table props

- now detected automatically in `columns`, there's no need to pass this prop manually

cee-chen committed:
Remove `hasActions` prop + util

- as far as I can tell this util isn't used anywhere else, so I'm deleting it entirely

cee-chen committed:
Remove `isExpandable` table prop

- this is now automatically detected by the presence of the `itemIdToExpandedRowMap` prop and no longer needs to be passed manually

cee-chen committed:
Remove `isExpandable` table prop with conditional logic

- was confused about this as the condition for `itemIdToExpandedRowMap` wasn't the same as `isExpandable`, so I updated the row map to support both

cee-chen committed:
Remove `isSelectable` table prop

- this is now automatically detected by the presence of the `selectable` prop and no longer needs to be passed manually

cee-chen committed:
Remove `isSelectable` conditional table props

- basic validation against `selection` conditionals, if they existed

- either way, this only affects responsive table views

+ remove unit test that no longer applies

cee-chen committed:
[EuiTableRow] `isSelectable` prop update

- `isSelectable` now indicates whether the checkbox is disabled or enabled; `hasSelection` (new prop) determines visual checkbox affordance

cee-chen committed:
[EuiTables] Misc syntax cleanups/chores

- remove unnecessary props that already default to false

- various syntax nits

- small CSS fixes

cee-chen committed:
Remove direct references to table classNames

- they no longer have styles attached to them and thus no longer do anything

- replace with direct component usage instead (or in some cases remove if they're not doing anything)

cee-chen committed:
Remove deprecated `showOnHover` column API and direct className usage

- replace with custom CSS instead for now, with a comment

cee-chen committed:
[Security] Replace role tables' custom actions/CSS with default actions

- now that EuiBasicTable's default actions supports passing the item to `href` as well as the click event to `onClick`

cee-chen commented:
/ci

cee-chen commented:
/ci

cee-chen committed:
Update DOM snapshots of tables

cee-chen commented:
/ci

cee-chen commented:
/ci

cee-chen commented:
/ci

cee-chen committed:
Update Enzyme snapshots

cee-chen committed:
Fix Enzyme selectors

- Emotion adds a wrapper with the same props, so we need to specify the actual DOM node

cee-chen committed:
Fix test text assertions caused by mobile headers no longer rendering to the DOM on desktop

cee-chen commented:
/ci

cee-chen committed:
Fix FTR selectors around actions

`findDescendent` requires "displayed" state, aka for visibility to be true, which in this case isn't because of `showOnHover` action opacity CSS/UX

Switching to just `.find` instead doesn't have a visibility requirement

cee-chen committed:
Fix Enzyme simulated events

cee-chen committed:
Fix/simplify more mobile vs desktop test assertions

- DOM no longer renders for mobile vs desktop whatsoever, so we get to remove a bunch of test utils and just mock a mobile window width instead

cee-chen commented:
/ci

cee-chen committed:
Fix failing FTR tests caused by examples missing EUI breakpoints provider

- without the provider, tables will incorrectly rendering in mobile view

cee-chen commented:
/ci

cee-chen committed:
Update table cell chart CSS to account for changes

- `className`/`css` not applies directly to the `<td>` cell and not the content

cee-chen committed:
Fix FTR selectors to account for removed modifier classNames/updated DOM

cee-chen commented:
/ci

cee-chen commented:
/ci

cee-chen commented:
/ci

cee-chen commented on x-pack/plugins/observability_solution/infra/public/components/asset_details/tabs/metadata/add_pin_to_row.tsx:
```diff
@@ -58,8 +61,22 @@ export const AddMetadataPinToRow = ({
+
   return (
-    <span className="euiTableCellContent__hoverItem expandedItemActions__completelyHide">
+    <span className={showOnRowHoverCss}>
```
@jennypavlova @elastic/infra-monitoring-ui I'd appreciate your help QAing this and confirming the hover UI/UX still works as expected. We've removed the above listed `className`s and the `showOnHover` API generally from columns, so this had to be converted to custom CSS instead.

cee-chen commented:
/ci

cee-chen commented on x-pack/plugins/security/public/management/roles/roles_grid/roles_grid_page.tsx:
@XavierM @elastic/security I thought I'd finally implement the EUI changes/overrides discussed in https://github.com/elastic/kibana/pull/118434#discussion_r749691551 :) 

I'd appreciate help fully QAing this to make sure the actions for roles and role mappings behave as before with no regressions.

There are some expected minor visual/UI changes that brings this table more in line with standard/existing EuiBasicTable actions, but other than that the actual behavior of clicking to edit/clone/delete roles should work as before.

<img width="152" alt="" src="https://github.com/elastic/kibana/assets/549407/46b5bf52-9ca0-4bab-b379-775551cbb9ac">

note: the `Delete` button losing its color is probably the most major impact. If requested, I can look into allowing `color` to properly apply to collapsed actions on EUI's side of things.

elasticmachine commented:
Pinging @elastic/eui-team (EUI)

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

stratoula commented:
I see that we are merging this in 8.15.0 😅 Very good decision, seems as a big change to do before the FF 😄

jbudz reviewed (approved):
`packages/kbn-test-jest-helpers/src/testbed/testbed.ts`, `src/dev/license_checker/config.ts`

seanrathier reviewed (approved):
Cloud Security Posture tested and approved.

jillguyonnet reviewed (approved):
Fleet changes LGTM ✅

nreese reviewed (commented):
Dashboard listing page looks odd when a description is provided. Notice how the name does not line up on the left.

<img width="500" alt="Screenshot 2024-04-12 at 9 07 49 AM" src="https://github.com/elastic/kibana/assets/373691/740887b3-4787-4e8d-bd43-a4b141908301">

cee-chen commented:
Thanks a million for catching that Nathan! I'll push out a fix/patch release for that!

SiddharthMantri reviewed (approved):
Security plugin changes tested. I've also confirmed with @elastic/kibana-security team that we're okay with the text color change in the Delete button. 🚀

dmlemeshko reviewed (approved):
`@kbn/test-jest-helpers` changes lgtm

cee-chen committed:
Merge remote-tracking branch 'upstream/main' into eui/v94.1.0

cee-chen committed:
Update EUI to grab backport with latest table fixes

+ more i18n updates from a11y improvements

cee-chen commented:
@nreese This should now be fixed if you pull the latest from this branch! Thanks again for catching this!

<img width="427" alt="" src="https://github.com/elastic/kibana/assets/549407/397bdeca-693c-480a-b1d0-1e9cf03c5d08">

nreese reviewed (approved):
kibana-presentation changes LGTM
code review, tested in chrome

vitaliidm reviewed (approved):
Detection engine area LGTM
Smoke tested exceptions tables

umbopepato reviewed (approved):
`triggers_actions_ui` changes LGTM!

jennypavlova reviewed (commented):
Great work 👏  I tested the infra and APM tables and almost everything works fine. I found some small issues I am listing here: 
- Infra > Hosts view or Inventory > Hosts details flyout or page > Processes Tab
(to see the host details click on any host name in the table and click on the processes tab)
![image](https://github.com/elastic/kibana/assets/14139027/46bd9a06-dbe0-4588-bccc-0910807a99e9)
It's the same on full screen:
![image](https://github.com/elastic/kibana/assets/14139027/252935cd-4312-4f92-bf2d-1a809b0be872)

- Infra > Hosts view
   - Maybe not really a bug but the header titles of the table on smaller screens are not visible because the arrows icon is now always present (Can we think about a way to hide it at least for small screen size? - the table looks good on a bigger screen): 
    ![image](https://github.com/elastic/kibana/assets/14139027/d947a4cb-1f36-45d5-8231-9e61a91f44c6)

- APM > Service Overview - The space around the table row is super big (Transactions and Dependencies tables)
   - Transactions
     ![image](https://github.com/elastic/kibana/assets/14139027/388e4544-cc08-40ad-a988-77bfd9874af3)
    - Dependencies
      <img width="1919" alt="image" src="https://github.com/elastic/kibana/assets/14139027/a1fc5636-2dca-43be-a599-eeb9f65db897">

PhilippeOberti reviewed (approved):
Changes LGTM for the Threat Hunting Investigations team. These are welcome simplifications and changes, thanks EUI team!! <3

jughosta reviewed (commented):
Hi @cee-chen,

Thanks for the update! Left a comment about an issue on Discover page.

maximpn reviewed (approved):
Rules Management Area LGTM

Smoked tested Rules Management, Rules Monitoring, Install Prebuilt rules, Update Prebult rules, ML Popover and Execution log table tables.

yuliacech reviewed (approved):
Kibana Management changes LGTM, tested the plugins locally and all looked okay

cee-chen commented on src/plugins/unified_doc_viewer/public/components/doc_viewer_table/legacy/table.tsx:
```diff
@@ -116,8 +116,7 @@ export const DocViewerLegacyTable = ({
       rowProps={onSetRowProps}
-      pagination={false}
-      responsive={false}
+      responsiveBreakpoint={false}
```
Ah thanks for flagging that, I added flex wrapping CSS to actions in the refactors to prevent overflow. You can either override that CSS or `flex-wrap: nowrap` or increase the width. If no objections I'll increase the width in this PR!

jughosta replied:
Increasing the width sounds good to me, thanks!

cee-chen replied:
https://github.com/elastic/kibana/pull/180514/commits/8d9b3627a7609bc13fa8283984e5672a2a0d6b36

jughosta replied:
Looking at this again I think the problem could be in the larger cell paddings than before. Also icons and text is now misaligned. 
<img width="500" alt="Screenshot 2024-04-16 at 20 10 18" src="https://github.com/elastic/kibana/assets/1415710/01a42d18-76fa-49c0-989f-7c81f56b2145">

I can work on this tomorrow as it should not go to any serverless deploy.

cee-chen replied:
Ahhh okay that's a custom cell CSS override in that case (see PR description). I've fixed it in https://github.com/elastic/kibana/pull/180514/commits/4055eff2c627bb8fee1bee32d8e21a4425b245de, thank you for flagging that!

jughosta replied:
Perfect, thanks!

cee-chen committed:
Merge branch 'main' into eui/v94.1.0

cee-chen committed:
[PR feedback] Increase actions column width

cee-chen committed:
[PR feedback] Fix custom table cell `className` CSS

jughosta reviewed (approved):
Data Discovery changes LGTM 👍

MichaelMarcialis reviewed (approved):
Thanks, @cee-chen! Looks good from my perspective. Approving.

@elastic/kibana-management: In testing, I noticed that the "Type" column is now truncated by default on the saved object page in stack management. This is due to the new addition of the `sortable` icon that appends each sortable column header. Ya'll may want to increase the width of this column slightly so that the text isn't always truncated.

![CleanShot 2024-04-17 at 10 32 46](https://github.com/elastic/kibana/assets/3884767/2b07709a-3210-498a-a3e6-db7e6f2b66a3)

cee-chen committed:
Merge branch 'main' into eui/v94.1.0

cee-chen committed:
[Infra][Hosts] Fix truncation on `Command` column

cee-chen committed:
[APM][Services] Fix overview tables custom CSS

- `:first-child` CSS no longer works well since `EuiBasicTable` no longer adds a `<div>` wrapper around the `<table>`

+ additionally fix Errors table to align the pagination bar to the bottom, like the Transactions and Dependencies tables

cee-chen committed:
Increase various column widths with text that are now cut off due to new sortable icon

+ use first custom `responsiveBreakpoint` to handle table with lots of columns!! 🎉

cee-chen commented:
@MichaelMarcialis I went ahead and fixed that in this PR - let me know if this looks good to you, and also if you find any other similar issues!

<img width="365" alt="" src="https://github.com/elastic/kibana/assets/549407/1aa24b5f-509e-4433-911b-7f915306cf95">

@jennypavlova Thank you so much for the help in getting Obs data into my local, I should have fixed all the issues in the screenshots in https://github.com/elastic/kibana/pull/180514#pullrequestreview-2001666042. I additionally fixed a small visual bug alreaedy in prod where the `Errors` table pagination wasn't aligned the same way that the Transactions and Dependencies tables were.

Also the Hosts overview table is now the first table to use the new `responsiveBreakpoint` feature - it has a lot of columns and text was cutting off very fast even on prod, so I changed it to collapse down to cards much sooner. Let me know what you think! 🎉

cee-chen commented:
👋 

@elastic/security-entity-analytics 
@elastic/security-defend-workflows
@elastic/security-threat-hunting-explore
@elastic/security-generative-ai
@elastic/ml-ui
@elastic/obs-ux-logs-team
@elastic/appex-sharedux
@elastic/kibana-esql
@elastic/enterprise-search-frontend
@elastic/obs-ai-assistant
@elastic/logstash
@elastic/kibana-visualizations 

One last ping for QA and review requests by EOD Thursday - we'll be asking for an admin merge if not. As always, even if you catch any visual bugs post-merge, the EUI team is more than happy to help quickly fix any issues after the fact.

mbondyra reviewed (approved):
Tested usage of EuiBasicTable for Lens. LGTM 👌🏼

jennypavlova reviewed (approved):
LGTM 🚀 @cee-chen Thank you very much for all the fixes! I tested it and all the tables look good ✨

semd reviewed (approved):
Security Threat Hunting Explore - Relevant usages of `EuiTable` tested, all good.
LGTM

semd commented:
I went a bit too fast with the approval, I just realized we have an issue very similar to [this one](https://github.com/elastic/kibana/pull/180514#pullrequestreview-2006237456) in our Data Quality Dashboard page. 

<img width="845" alt="results_column_issue" src="https://github.com/elastic/kibana/assets/17747913/73497f1f-4a14-49f5-88c6-eb4c5567862f">

We only need to increase the width here:

https://github.com/elastic/kibana/blob/d3207ec1ec0fb9b6cc43dedf0544e85bf4f90a48/x-pack/packages/security-solution/ecs_data_quality_dashboard/impl/data_quality/data_quality_panel/summary_table/helpers.tsx#L197

To `70px`:

<img width="845" alt="result_column_fixed" src="https://github.com/elastic/kibana/assets/17747913/3a51e214-dd70-4425-813f-74b008c7e39a">


@cee-chen can you apply the change in this PR?

cee-chen committed:
Update snapshot from updated table

cee-chen committed:
[Security][Data Quality] Increase width of column to account for sortable icon

cee-chen committed:
Merge branch 'main' into eui/v94.1.0

cee-chen commented:
@semd Updated! Thank you for catching that!

sphilipse reviewed (approved):
smoke-tested Search changes, LGTM!

cee-chen committed:
Update test w/ updated column width
